src/D3MJob.sol

```diff
@@ -0,0 +1,126 @@
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+pragma solidity 0.8.13;
+
+import {IJob} from "./interfaces/IJob.sol";
+
+interface SequencerLike {
+    function isMaster(bytes32 network) external view returns (bool);
+}
+
+interface IlkRegistryLike {
+    function list() external view returns (bytes32[] memory);
+}
+
+interface VatLike {
+    function urns(bytes32, address) external view returns (uint256, uint256);
+}
+
+interface D3MHubLike {
+    function vat() external view returns (VatLike);
+    function pool(bytes32) external view returns (address);
+    function exec(bytes32) external;
+}
+
+/// @title Trigger D3M updates based on threshold
+contract D3MJob is IJob {
+
+    uint256 constant internal BPS = 10 ** 4;
+    
+    SequencerLike public immutable sequencer;
+    IlkRegistryLike public immutable ilkRegistry;
+    D3MHubLike public immutable hub;
+    VatLike public immutable vat;
+    uint256 public immutable threshold;             // Threshold deviation to kick off exec [BPS]
+    uint256 public immutable ttl;                   // Cooldown before you can call exec again [seconds]
+
+    mapping (bytes32 => uint256) public expiry;     // Timestamp of when exec is allowed again
+
+    // --- Errors ---
+    error NotMaster(bytes32 network);
+    error Cooldown(bytes32 ilk, uint256 expiry);
+    error ShouldNotTrigger(bytes32 ilk, uint256 part, uint256 nart);
+
+    // --- Events ---
+    event Work(bytes32 indexed network);
+
+    constructor(address _sequencer, address _ilkRegistry, address _hub, uint256 _threshold, uint256 _ttl) {
+        sequencer = SequencerLike(_sequencer);
+        ilkRegistry = IlkRegistryLike(_ilkRegistry);
+        hub = D3MHubLike(_hub);
+        vat = hub.vat();
+        threshold = _threshold;
+        ttl = _ttl;
+    }
+
+    function shouldTrigger(uint256 part, uint256 nart) internal view returns (bool) {
+        if (part == 0 && nart != 0) return true;    // From zero to non-zero
+        if (part != 0 && nart == 0) return true;    // From non-zero to zero
+
+        // Check if the delta is above the threshold
+        uint256 delta = nart * BPS / part;
+        if (delta < BPS) delta = BPS * BPS / delta; // Flip decreases to increase
+
+        return delta >= (BPS + threshold);
+    }
+
+    function work(bytes32 network, bytes calldata args) external override {
+        if (!sequencer.isMaster(network)) revert NotMaster(network);
+
+        bytes32 ilk = abi.decode(args, (bytes32));
+        uint256 _expiry = expiry[ilk];
+        if (block.timestamp < _expiry) revert Cooldown(ilk, _expiry);
+        address pool = hub.pool(ilk);
+        (, uint256 part) = vat.urns(ilk, pool);
+
+        hub.exec(ilk);
+
+        (, uint256 nart) = vat.urns(ilk, pool);
+        if (!shouldTrigger(part, nart)) revert ShouldNotTrigger(ilk, part, nart);
```

Comment:
I'm not super excited about how expensive this is.  I get that this can be tested for success before execution, but one could just call workable too.  If someone calls the `exec()` before you, you still have to eat the gas as though you called it.  All this saves is the expiry math and the `Work()` event.

Unless I'm missing something, I would remove this `shouldTrigger()` check from `work()` and all the `SLOAD`s it requires.